tests/api_connexion/schemas/test_role_and_permission_schema.py

```diff
@@ -0,0 +1,104 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import unittest
+
+from airflow.api_connexion.schemas.role_and_permission_schema import (
+    RoleCollection,
+    role_collection_item_schema,
+    role_collection_schema,
+)
+from airflow.security import permissions
+from airflow.www import app
+from tests.test_utils.api_connexion_utils import create_role, delete_role
+
+
+class TestRoleCollectionItemSchema(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.app = app.create_app(testing=True)  # type:ignore
+        cls.role = create_role(
+            cls.app,  # type: ignore
+            name="Test",
+            permissions=[
+                (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
+            ],
+        )
+
+    def tearDown(self) -> None:
+        delete_role(self.app, 'Test')
+
+    def test_serialize(self):
+        deserialized_role = role_collection_item_schema.dump(self.role)
+        assert deserialized_role == {
+            'name': 'Test',
+            'actions': [{'resource': {'name': 'Connections'}, 'action': {'name': 'can_create'}}],
+        }
+
+    def test_deserialize(self):
+        role = {
+            'name': 'Test',
+            'actions': [{'resource': {'name': 'Connections'}, 'action': {'name': 'can_create'}}],
+        }
+        role_obj = role_collection_item_schema.load(role)
+        assert role_obj == {
+            'name': 'Test',
+            'permissions': [{'view_menu': {'name': 'Connections'}, 'permission': {'name': 'can_create'}}],
+        }
+
+
+class TestRoleCollectionSchema(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.app = app.create_app(testing=True)  # type:ignore
+        cls.role1 = create_role(
+            cls.app,  # type: ignore
+            name="Test1",
+            permissions=[
+                (permissions.ACTION_CAN_CREATE, permissions.RESOURCE_CONNECTION),
+            ],
+        )
+        cls.role2 = create_role(
+            cls.app,  # type: ignore
+            name="Test2",
+            permissions=[
+                (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG),
+            ],
+        )
+
+    def tearDown(self) -> None:
+        delete_role(self.app, 'Test1')
+        delete_role(self.app, 'Test2')
+
+    def test_serialize(self):
+        instance = RoleCollection([self.role1, self.role2], total_entries=2)
+        deserialized = role_collection_schema.dump(instance)
+        assert deserialized == {
+            'roles': [
+                {
+                    'name': 'Test1',
+                    'actions': [{'resource': {'name': 'Connections'}, 'action': {'name': 'can_create'}}],
```

Comment:
Hmmmm. I wonder if this should  instead be

```
                    'actions': [{'resource': 'Connections', 'action': 'can_create'}],
```

It's clearer/simpler right now, I guess the question is if we might want to include any other properties in here in the future.

WDYT @mik-laj @jhtimmins ?

Reply:
I first serialized it that way but I couldn't deserialize it which would pose a problem during Role creation with permissions

Reply:
I prefer to use objects to make this API more future-proof.  This may also make it easier to use this API, despite appearances, because in strongly typed languages you will have a better type hint. You will be able to go to the class documentation and not just the description of the attribute.